### ISSUE TITLE
G1 canary agreement is 52% vs the ≥95% floor: rust-sort/explain-reverse diverge at token 0

### ISSUE THREAD

**Issue by gpillon** [gpillon/ignis#72]
## Context

Master #36 / gate G1 (#62 records the verdict). `.scratch/PENDING.md` already describes this as an open, unticketed blocker — filing it as a tracked issue per #62's acceptance criterion ("any gap found is filed as a new ticket rather than waived").

## Evidence (as recorded in PENDING.md, post RMSNorm `unit_offset` fix #67, thinking disabled to match how the oracle was recorded)

- `rust-hello`: 21/21 — exact agreement with the reference.
- `math-greedy`: 32/32 — exact agreement with the reference.
- `rust-sort`: 0/32 — diverges at token 0.
- `explain-reverse`: 0/17 — diverges at token 0.
- Suite average: 52%, against the ≥95% G1 floor (`ignis-bench oracle compare` vs `crates/bench/tests/fixtures/oracle_canary.json`).

Both divergent prompts still produce correct, fluent answers (`` `v` is set to `[1, 2, 3]`. `` and a correct one-sentence description of `Vec::reverse`) — just different wording from the oracle's. This is not the #67 corruption class of failure. Working theory: an argmax near-tie flipped by a residual numeric difference; once flipped, the trajectory diverges entirely.

## What's needed to confirm

Logit-level comparison: top-k token ids + values after the final head, same tokenized prompt, both engines, at the point of first divergence. The server does not expose logits today. PENDING.md notes this needed `enable_thinking` (#68) to reproduce the oracle's prompt through the HTTP surface — #68 has since landed (commit ccc4fae), so this may now be unblocked; needs re-checking.

## Blocks

G1 gate (#62 / master #36): the acceptance criterion is ≥95% first-32-token agreement per canary prompt; current suite average is 52%.

**Comment by gpillon** [comment]
## Root cause confirmed: exact BF16 argmax ties, not "near"-ties

Added an `out_logits` escape hatch to the real 64-layer program path (`ignis_program_prefill`, mirroring the degenerate path's existing copy-back — `kernel/src/step.cu`, `crates/core/src/step.rs::prefill_program`) and a GPU test, `crates/server/tests/logit_divergence_gpu.rs`, that reproduces the oracle's exact prompt (`ArtifactTemplateProvider`, thinking disabled per #68) and dumps token 0's top-8 logits.

Run against the real artifact on a free RTX 5090:

```
rust-sort:       emitted=63  top8=[(63, 19.5), (5836, 19.5), (760, 18.875), (623, 18.25), (13962, 17.375), (2064, 17.25), (332, 17.125), (58, 16.375)]
                 oracle_first=5836 oracle_rank=1  gap = 0.0 (winner=19.5, oracle=19.5)

explain-reverse: emitted=760 top8=[(760, 22.875), (2064, 22.875), (63, 22.625), (623, 20.0), (47218, 18.125), (674, 17.625), (49, 16.875), (71093, 16.75)]
                 oracle_first=2064 oracle_rank=1  gap = 0.0 (winner=22.875, oracle=22.875)
```

Both canaries hit a **bit-identical BF16 tie** between ignis's winner and the oracle's expected token at token 0 (not a small numeric gap — literally the same BF16 value). `ninfer::ops::argmax`'s lowest-index tie-break picks 63/760 over the oracle's 5836/2064 in both cases. This is not a forward-pass correctness bug — the per-layer output already matches the f64 reference within BF16 tolerance (`step_degenerate_gpu.rs`, `gqa_layer_gpu.rs`, `gdn_layer_gpu.rs`, all green on this run) — it's the model being genuinely ambivalent at BF16 resolution between two next tokens, with the reference engine's tie-break (or higher-precision output head) landing on the other one.

**Still open, not resolved by this ticket:** what to do about it. Options: match the reference's tie-break convention (not confirmed — only inferred from these two outcomes), compute the final head at higher precision to avoid the tie in the first place, or treat this as inherent BF16 rounding noise and revisit the G1 ≥95% floor / canary set. That's a gate-policy decision, not a code fix, so leaving this open rather than closing it here.

Full GPU profile (`scripts/gpu-profile.ps1`) is green end to end with this change, including every pre-existing GPU test — no regressions.

**Comment by gpillon** [comment]
Root cause confirmed on a free RTX 5090 (2026-09-07): this is an **exact BF16 logit tie**, not just a "near" one.

`ignis_program_prefill` (kernel/src/step.cu, kernel/include/ignis_step.h) now optionally returns the prompt span's last-position logits — the same copy-back pattern the degenerate/test-only path already used, extended to the real 64-layer program. `crates/core/src/step.rs::prefill_program` threads it through. A new GPU test, `crates/server/tests/logit_divergence_gpu.rs`, replays `rust-sort` and `explain-reverse` through the real chat template with thinking disabled (matching how the oracle fixture was recorded) and reports the top-8 logits at token 0, sourced straight from `crates/bench/tests/fixtures/oracle_canary.json` so the prompt/expected-token can't drift from what the gate actually scores.

Results, run against the real artifact:

- `rust-sort`: top-2 are token 63 (ignis's emitted token) and token 5836 (the oracle's expected token), **both at logit 19.5**.
- `explain-reverse`: top-2 are token 760 (emitted) and token 2064 (oracle's expected), **both at logit 22.875**.

ignis's argmax breaks ties toward the lowest token id (`ninfer::ops::argmax`'s documented convention), which is exactly why it picks 63/760 instead of the oracle's 5836/2064. Both continuations are correct, fluent completions — consistent with the issue's original observation that this isn't the #67 corruption class of failure. The model is genuinely ambivalent at BF16 precision between two essentially-equal next tokens; the reference engine's own tie-break (or a higher-precision output head) landed on the other member of the tie.

Note on scope: only ignis's own logits were instrumented here — the reference engine is a separate codebase and wasn't touched. We didn't extract its internal logit values, but ignis's runner-up being *exactly* the oracle's chosen token at an exact tied logit is strong evidence on its own.

This isn't a forward-pass bug, so there's nothing further to "fix" in the compute path. What's still open is a **gate-policy decision**: how should G1's ≥95% first-32-token-agreement floor treat a BF16-exact-tie divergence? Options include matching the reference's tie-break convention, widening the gate to tolerate ties, or accepting that this metric tops out below 95% against a different engine's output-head rounding. Recorded in `.scratch/PENDING.md`; leaving this issue open pending that call.

**Comment by gpillon** [comment]
Merged to main (0993d1f) — GitHub auto-closed this on push (commit message keyword). Diagnosis complete, see the comment above for the confirmed root cause (exact BF16 logit tie, not a forward-pass bug). The remaining open question (how the G1 gate should treat a tie-break divergence) is a policy decision, tracked separately in #76.
